api/v1/model_types.go

```diff
@@ -6,47 +6,40 @@ import (
 
 // ModelSpec defines the desired state of Model
 type ModelSpec struct {
-	// Source is a reference to the source of the model.
-	Source ModelSource `json:"source"`
-	// Training should be set to run a training job.
-	Training *ModelTraining `json:"training,omitempty"`
-	// Size describes different size dimensions of the underlying model.
-	Size ModelSize `json:"size"`
-	// Compute describes the compute requirements and preferences of the model.
-	Compute ModelCompute `json:"compute"`
-}
+	// Container that contains model code and dependencies.
+	Container Container `json:"container"`
+
+	// Loader should be set to run a loading job.
+	Loader *ModelLoader `json:"trainer,omitempty"`
 
-type ModelCompute struct {
-	// +kubebuilder:validation:MinItems=1
-	// Types is a list of supported compute types for this Model. This list should be
-	// ordered by preference, with the most preferred type first.
-	Types []ComputeType `json:"types"`
+	// Trainer should be set to run a training job.
+	Trainer *ModelTrainer `json:"trainer,omitempty"`
 }
 
-// +kubebuilder:validation:Enum=CPU;GPU
-type ComputeType string
+func (m *Model) GetContainer() Container {
+	return m.Spec.Container
+}
+func (m *Model) SetContainer(c Container) {
+	m.Spec.Container = c
+}
 
-const (
-	ComputeTypeCPU ComputeType = "CPU"
-	ComputeTypeGPU ComputeType = "GPU"
-	//ComputeTypeTPU ComputeType = "TPU"
-)
+func (m *Model) GetConditions() *[]metav1.Condition {
+	return &m.Status.Conditions
+}
 
-type ModelSize struct {
-	// ParameterCount is the number of parameters in the underlying model.
-	ParameterCount int64 `json:"parameterCount,omitempty"`
-	// ParameterBits is the number of bits per parameter in the underlying model. Common values would be 8, 16, 32.
-	ParameterBits int `json:"parameterBits,omitempty"`
+type ModelLoader struct {
+	Params map[string]string `json:"params"`
 }
 
-type ModelTraining struct {
-	// DatasetName is the .metadata.name of the Dataset to use for training.
-	DatasetName string `json:"datasetName"`
+type ModelTrainer struct {
+	// Dataset to mount for training.
+	Dataset ObjectRef `json:"datasetName"`
+
 	// Params is a list of hyperparameters to use for training.
-	Params ModelTrainingParams `json:"params"`
+	Params ModelTrainerParams `json:"params"`
```

Comment:
I thought about this... What about well-known params like epochs? I wonder if it makes sense to have these strongly-typed and a part of the API schema? This would encourage the use of the standard params, which would help with things like automated hyperparameter tuning down the road.

Reply:
What if some end-user needs to pass more params to their training? My take is we should move any strongly typed well known params 1 level up and leave Params to be dynamic in nature.

Reply:
I think thats a good call, either that or have `.params` and `.customParams`.

Reply:
I prefer to error on flexibility here instead of requiring further API changes down the line.

Reply:
I think we should start flexible and then back out the most common hyperparameters into strongly typed fields as they emerge. Maybe something like:

```suggestion
type ModelTrainer struct {
	// Dataset to mount for training.
	Dataset ObjectRef `json:"datasetName"`

	// Params is a list of hyperparameters to use for training.
	Params ModelTrainerParams `json:"params"`
}

type ModelTrainerParams struct {
	//+kubebuilder:default:=3
	// Epochs is the total number of iterations that should be run through the training data.
	// Increasing this number will increase training time.
	BatchSize int64 `json:"batchSize,omitempty"`
	Custom map[string]interface{} `json:"custom,omitempty"`
}
```

Reply:
my preference is to not have any strongly types params to improve flexibility and support all kinds of training jobs. Any param should be optional and have a default.

Reply:
Agreed on defaults

Reply:
I like the proposal... I currently have `.custom` implemented as `.params` and moved the standard ones (`.epoch`) up to the `.trainer` level... I like having all hyperparams organized under `.params` though.

Reply:
@samos123 If we dont standardize some common hyperparameters as typed values, I worry that we will miss out on hyperparameter tuning functionality in the future.

Reply:
> @samos123 If we dont standardize some common hyperparameters as typed values, I worry that we will miss out on hyperparameter tuning functionality in the future.

If we start untyped, we're still have a non-breaking API change away from adding typed hyperparams. I realize we're very pre-1.0 but the opposite direction would be API breaking.

Reply:
You want hyperparameters? then use our standard naming scheme otherwise miss out on a feature.  I just don't like enforcing parameters in the API if that limits potential future use cases. I think it's too early to start enforcing parameters. Any parameter (including epochs)  should be optional and can be ignored by a trainer container.